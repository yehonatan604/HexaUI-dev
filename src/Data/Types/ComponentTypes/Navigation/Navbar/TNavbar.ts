import { TComponent, ComponentOptions } from "../../TComponent";

export type TNavbar = TComponent<"div"> & {
    options?: ComponentOptions;
}
