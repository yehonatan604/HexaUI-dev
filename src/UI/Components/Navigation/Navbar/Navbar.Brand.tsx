import useTheme from "../../../../Core/Context/ThemeContext/hooks/useTheme";
import { TBrand } from "../../../../Data/Types/TBrand";

const NavbarBrand = (props: TBrand) => {
  const { children, as, to, options, ...componentProps } = props;
  const { colors, mode } = useTheme();

  // Options
  const brandColor =
    options?.textVariant || (mode === "light" ? colors.secondary : colors.whiteDark);
  const bgColor = options?.bgVariant || "transparent";

  const Component = as || "a";

  // JSX
  return (
    <Component
      className={`text-xl font-bold text-${brandColor} bg-${bgColor}`}
      {...componentProps}
      href={to}
      to={to}
    >
      {children}
    </Component>
  );
};

export default NavbarBrand;
