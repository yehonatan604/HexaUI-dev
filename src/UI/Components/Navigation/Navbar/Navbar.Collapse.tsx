import useTheme from "../../../../Core/Context/ThemeContext/hooks/useTheme";
import { FlexDir } from "../../../../Data/Constants/FlexDirection";
import { FlexTypes } from "../../../../Data/Constants/FlexTypes";
import { TNavbarCollapse } from "./types/TNavbarCollapse";
import Flex from "../../Layout/Flex/Flex";

const NavbarCollapse = (props: TNavbarCollapse) => {
  const { children, className, isMobileMenuOpen, options, ...componentProps } = props;

  const { colors, mode } = useTheme();

  const textColor =
    options?.textVariant || (mode === "light" ? colors.secondary : colors.standradLight);

  // JSX
  return (
    <Flex
      options={{
        direction: isMobileMenuOpen ? FlexDir.Col : FlexDir.Row,
        align: FlexTypes.Center,
      }}
      className={`text-${textColor} flex ${
        isMobileMenuOpen ? "md:flex-row" : "max-md:hidden"
      } gap-4 ${className}`}
      {...componentProps}
    >
      {children}
    </Flex>
  );
};

export default NavbarCollapse;
