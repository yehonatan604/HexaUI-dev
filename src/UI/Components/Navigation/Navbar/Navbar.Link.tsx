import { Link, useLocation } from "react-router-dom";
import { TNavbarLink } from "./types/TNavbarLink";
import useTheme from "../../../../Core/Context/ThemeContext/hooks/useTheme";

const NavbarLink = (props: TNavbarLink) => {
  const { children, options, to, ...componentProps } = props;

  // Color Palette
  const { colors, mode } = useTheme();

  // Options
  const path = useLocation().pathname.split("/")[1];
  const isActive = path === to.split("/")[1];
  const textColor =
    options?.textVariant || (mode === "light" ? colors.secondary : colors.whiteDark);
  const activeLinkColor = options?.activeVariant || colors.warning;
  const linkColor = isActive ? activeLinkColor : textColor;

  // JSX
  return (
    <Link
      className={`text-lg text-${linkColor} font-semibold hover:opacity-70`}
      {...componentProps}
      to={to}
    >
      {children}
    </Link>
  );
};

export default NavbarLink;
