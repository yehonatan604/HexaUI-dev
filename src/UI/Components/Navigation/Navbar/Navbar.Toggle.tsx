import { TNavbarToggle } from "./types/TNavbarToggle";

const NavbarToggle = (props: TNavbarToggle) => {
  const { children, toggle, ...componentProps } = props;

  const onToggle = () => {
    toggle((prev) => !prev);
  };
  return (
    <button className="md:hidden" {...componentProps} onClick={onToggle}>
      {children}
    </button>
  );
};

export default NavbarToggle;
