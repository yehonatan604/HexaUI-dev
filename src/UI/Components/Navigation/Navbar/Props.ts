import { TComponent } from "../../../../Data/Types/ComponentTypes/TComponent";

export type NavbarProps = TComponent<"div"> & {
    variant?: string;
    colors?: {
        lightColor?: string;
        darkColor?: string;
    };
};