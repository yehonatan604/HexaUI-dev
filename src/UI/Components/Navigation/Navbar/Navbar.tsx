import useTheme from "../../../../Core/Context/ThemeContext/hooks/useTheme";
import { FlexTypes } from "../../../../Data/Constants/FlexTypes";
import { TNavbar } from "./types/TNavbar";
import Flex from "../../Layout/Flex/Flex";
import NavbarBrand from "./Navbar.Brand";
import NavbarCollapse from "./Navbar.Collapse";
import NavbarLink from "./Navbar.Link";
import NavbarToggle from "./Navbar.Toggle";

const Navbar = (props: TNavbar) => {
  const { children, options, className, ...componentProps } = props;

  // Hooks
  const { colors } = useTheme();

  // Options
  const bgColor = options?.bgVariant || colors.primary;
  const textColor = options?.textVariant || colors.white;

  // JSX
  return (
    <Flex
      options={{
        justify: FlexTypes.SpaceBetween,
        align: FlexTypes.Center,
      }}
      {...componentProps}
      className={`bg-${bgColor} text-${textColor} p-4 max-md:flex-col max-md:gap-4 md:flex-row ${className}`}
    >
      {children}
    </Flex>
  );
};

Navbar.Toggle = NavbarToggle;
Navbar.Brand = NavbarBrand;
Navbar.Collapse = NavbarCollapse;
Navbar.Link = NavbarLink;

export default Navbar;
