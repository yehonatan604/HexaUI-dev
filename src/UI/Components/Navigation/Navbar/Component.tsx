import useColorPalette from "../../../../Core/Hooks/useColorPallete";
import useTheme from "../../../../Core/Hooks/useTheme";
import { FlexTypes } from "../../../../Data/Constants/FlexTypes";
import { Variants } from "../../../../Data/Constants/Variants";
import Flex from "../../Layout/Flex/Component";
import NavbarBrand from "./Navbar.Brand/Component";
import NavbarCollapse from "./Navbar.Collapse/Component";
import NavbarLink from "./Navbar.Link/Component";
import NavbarToggle from "./Navbar.Toggle/Component";
import { NavbarProps } from "./Props";

const Navbar = (props: NavbarProps) => {
  const { children, variant = Variants.Standard, colors, ...componentProps } = props;

  // Hooks
  const mode = useTheme().mode;
  const getColor = useColorPalette().getColor;

  // Color Palette
  const { light: lightStandard } = getColor(Variants.Standard)!;
  const { light: lightFailure } = getColor(Variants.Failure)!;
  const { light: lightSuccess } = getColor(Variants.Success)!;
  const { light: lightWarning } = getColor(Variants.Warning)!;
  const { light: lightInfo } = getColor(Variants.Info)!;
  const { light: lightLime } = getColor(Variants.Lime)!;
  const { light: lightCyan } = getColor(Variants.Cyan)!;
  const { light: lightPurple } = getColor(Variants.Purple)!;
  const { light: lightPink } = getColor(Variants.Pink)!;
  const { light: navbarTextColor } = getColor("navbarText")!;

  // Variables
  let bgColor: string = "";
  if (variant === Variants.Failure) {
    bgColor = `bg-${lightFailure}`;
  } else if (variant === Variants.Success) {
    bgColor = `bg-${lightSuccess}`;
  } else if (variant === Variants.Warning) {
    bgColor = `bg-${lightWarning}`;
  } else if (variant === Variants.Info) {
    bgColor = `bg-${lightInfo}`;
  } else if (variant === Variants.Lime) {
    bgColor = `bg-${lightLime}`;
  } else if (variant === Variants.Cyan) {
    bgColor = `bg-${lightCyan}`;
  } else if (variant === Variants.Purple) {
    bgColor = `bg-${lightPurple}`;
  } else if (variant === Variants.Pink) {
    bgColor = `bg-${lightPink}`;
  } else {
    bgColor = `bg-${lightStandard}`;
  }
  if (colors) bgColor = mode === "light" ? colors.lightColor : colors.darkColor;

  // JSX
  return (
    <Flex
      justify={FlexTypes.SpaceBetween}
      align={FlexTypes.Center}
      {...componentProps}
      className={`${bgColor} text-${navbarTextColor} p-4 max-md:flex-col max-md:gap-4 md:flex-row`}
    >
      {children}
    </Flex>
  );
};

Navbar.Toggle = NavbarToggle;
Navbar.Brand = NavbarBrand;
Navbar.Collapse = NavbarCollapse;
Navbar.Link = NavbarLink;

export default Navbar;
