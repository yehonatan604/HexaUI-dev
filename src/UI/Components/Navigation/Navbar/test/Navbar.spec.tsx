import { render, screen, fireEvent } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom"; // Needed for NavbarLink
import Navbar from "../Navbar"; // Adjust the import path
import { describe, it, expect, vi } from "vitest";

// Mock the useTheme hook
vi.mock("../../../../Core/Context/ThemeContext/hooks/useTheme", () => ({
  __esModule: true,
  default: () => ({
    colors: {
      primary: "blue-500",
      secondary: "gray-800",
      white: "white",
      warning: "yellow-500",
    },
    mode: "light",
  }),
}));

describe("Navbar Component", () => {
  it("should render children correctly", () => {
    render(
      <Router>
        <Navbar options={{ bgVariant: "blue-500" }}>
          <div>Child 1</div>
          <div>Child 2</div>
        </Navbar>
      </Router>
    );

    // Check if both children are rendered
    expect(screen.getByText("Child 1")).toBeInTheDocument();
    expect(screen.getByText("Child 2")).toBeInTheDocument();
  });

  it("should apply custom background and text colors", () => {
    render(
      <Router>
        <Navbar options={{ bgVariant: "blue-500", textVariant: "white" }}>
          <div>Navbar Content</div>
        </Navbar>
      </Router>
    );

    const navbarElement = screen.getByText("Navbar Content").parentElement;

    // Check if the custom background and text colors are applied
    expect(navbarElement).toHaveClass("bg-blue-500", "text-white");
  });

  it("should toggle mobile menu visibility on toggle click", () => {
    const toggleMock = vi.fn();
    render(
      <Router>
        <Navbar>
          <Navbar.Toggle toggle={toggleMock}>Toggle Menu</Navbar.Toggle>
          <Navbar.Collapse isMobileMenuOpen={true}>
            <div>Menu Content</div>
          </Navbar.Collapse>
        </Navbar>
      </Router>
    );

    const toggleButton = screen.getByText("Toggle Menu");

    // Check if the toggle button is rendered
    expect(toggleButton).toBeInTheDocument();

    // Simulate clicking the toggle button
    fireEvent.click(toggleButton);

    // Check if the toggle function is called
    expect(toggleMock).toHaveBeenCalled();
  });

  it("should highlight active NavbarLink when the route matches", () => {
    render(
      <Router>
        <Navbar>
          <Navbar.Link to="/home">Home</Navbar.Link>
        </Navbar>
      </Router>
    );

    const homeLink = screen.getByText("Home");

    // Check if the link is rendered
    expect(homeLink).toBeInTheDocument();

    // Check if the active link class is applied (default is secondary)
    expect(homeLink).toHaveClass("text-secondary");
  });

  it("should render and match the snapshot", () => {
    const { asFragment } = render(
      <Router>
        <Navbar
          options={{
            bgVariant: "blue-500",
            textVariant: "white",
          }}
        >
          <Navbar.Brand to="/">Brand</Navbar.Brand>
          <Navbar.Link to="/home">Home</Navbar.Link>
          <Navbar.Collapse isMobileMenuOpen={false}>
            <div>Menu Content</div>
          </Navbar.Collapse>
        </Navbar>
      </Router>
    );

    // Snapshot the Navbar structure
    expect(asFragment()).toMatchSnapshot();
  });
});
