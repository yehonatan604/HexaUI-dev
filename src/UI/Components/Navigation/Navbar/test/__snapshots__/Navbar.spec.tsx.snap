// Vitest Snapshot v1

exports[`Navbar Component > should render and match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-row bg- text- justify-between items-center flex-nowrap bg-blue-500 text-white p-4 max-md:flex-col max-md:gap-4 md:flex-row undefined"
  >
    <a
      class="text-xl font-bold text-secondary bg-transparent"
      href="/"
      to="/"
    >
      Brand
    </a>
    <a
      class="text-lg text-secondary font-semibold hover:opacity-70"
      href="/home"
    >
      Home
    </a>
    <div
      class="flex flex-row bg- text- justify-start items-center flex-nowrap text-secondary flex max-md:hidden gap-4 undefined"
    >
      <div>
        Menu Content
      </div>
    </div>
  </div>
</DocumentFragment>
`;
